src/algos/dtmscc.jl

```diff
@@ -0,0 +1,286 @@
+
+    while it<maxit && err>tol
+
+        it += 1
+        value_0 = copy(value)
+
+
+        fut_dr = create_interpolant(value_0)
+
+
+        for i_mc=1:n_mc
+            for n=1:N
+                s = slice(grid, n, :)
+                m = slice(nodes, i_mc, :)
+                x = policy(i_mc,s)
+                val = evaluate(felicity,m, s, x, p)
+                for j_mc in 1:n_mc
+                    prob = Q[i_mc, j_mc]
+                    # M = nodes[j_mc,:]
+                    M = slice(P, j_mc, :)
+                    S = evaluate(transition,m,s,x,M,p)
+                    v_fut = fut_dr[j_mc][S...]
+                    val += discount*prob*v_fut
+                end
+                value[i_mc, n:n] = val
+            end
+        end
+        err = maximum( abs(value - value_0)[:] )
+    end
+    return value
+end
+
+
+####
+# optimize policy rules by iterating on the value function
+####
+
+function solve_policy(model::DTMSCCModel, policy, values_0, options=Dict())
+
+    felicity = model.functions.felicity
+    transition = model.functions.transition
+
+    discount = model.calibration.flat[:beta]
+
+    # get approximation space
+    ap = model.options[:approximation_space]
+    a = ap[:a]
+    b = ap[:b]
+    orders = ap[:orders]
+    approx = ApproximationSpace(a,b,orders)
+
+    grid = mlinspace(a,b,orders)
+    d = size(grid,2)
+
+    # get markov chain
+    mcopts = model.options[:discrete_transition][:MarkovChain]
+    P = cat(1, [e' for e  in mcopts[1]]...)
+    Q = cat(1, [e' for e  in mcopts[2]]...)
+    markov_chain = MarkovChain(P,Q)
+    nodes = markov_chain.P
+
+    symbols = model.symbolic.symbols
+    calibration = model.calibration
+
+    N = size(grid,1)
+    n_mc = size(markov_chain.P,1)
+    n_s = size(grid,2)
+    n_x = length(symbols[:controls])
+    p = calibration
+    controls = zeros(n_mc,N,n_x)
+    value = zeros(n_mc,N)
+    p = calibration[:parameters]
+
+    controls = zeros(n_mc,N,n_x)
+    for i_mc=1:n_mc
+        for n=1:N
+            s = grid[n,:]
+            m = nodes[i_mc,:]
+            x = policy(i_mc,s)
+            controls[i_mc, n,:] = x
+        end
+    end
+
+    lower_bound = model.functions.controls_lb
+    upper_bound = model.functions.controls_ub
+
+#     controls = copy(controls)
+    value = copy(values_0)
+
+    it = 0
+    maxit = get(options, :maxit, 1000)
+    tol = get(options, :tol, 0.00001)
+    tol_2 = get(options, :tol_2, 0.00001)
+
+    diff = 100
+    diff_0 = 1.0
+    diff_2_0 = 1.0
+
+    dims = vcat(approx.orders)
+    knots = [collect(linspace(approx.a[i], approx.b[i], approx.orders[i])) for i=1:d]
+    knots = tuple(knots...)
+
+    # interpolations.jl does not have cubic gridded interpolations yet
+    # using splnes.jl isntead in the meantime
+     create_interpolant(vv) = [interpolant_cspline(approx.a, approx.b, approx.orders, copy(reshape(slice(vv,i_mc,:),dims...) )) for i_mc in 1:n_mc]
```

Comment:
Actually, I didn't really want to copy the array, but noticed that some of the functions called below didn't work with slices. Will do what you suggest.